experimental/generation/generator/packages/library/src/substitutions.ts

```diff
@@ -66,30 +71,37 @@ function substitutions(path: string, bindings: any, copies?: number, seed?: stri
     if (!seed) seed = '0'
 
     // Normalize bindings into a simple array and setup variables
-    let state = new Map<string, Variable>();
-    for (let binding of Object.keys(bindings)) {
-        let value = bindings[binding]
+    const state = new Map<string, Variable>();
+    for (const binding of Object.keys(bindings)) {
+        const value = bindings[binding]
         if (Array.isArray(value)) {
             state.set(binding, {index: -1, values: (value as any).flat()})
-        } else {
+        } else if (typeof value === 'string' || typeof value === 'number') {
             state.set(binding, {index: -1, values: value ? [value] : []})
         }
     }
 
-    let result: string[] = []
-    let rand = random(seed)
-    let file = fs.readFileSync(path, 'utf8')
+    // Track all generated utterances
+    if (!('utterances' in bindings)) {
+        // Setup memory of all generated strings
+        bindings.utterances = new Set<string>()
+    }
+    const utterances = bindings.utterances as Set<string>
```

Comment:
Why is the `as` casting necessary?

Reply:
Because bindings are not strongly typed.

---
In reply to: [535408040](https://github.com/microsoft/BotBuilder-Samples/pull/2967#discussion_r535408040) [](ancestors = 535408040)